Lib/multiprocessing/util.py

```diff
@@ -167,6 +171,165 @@ def _run_after_forkers():
+        # This should not happen in real life but can happen in
+        # test_concurrent_futures.
+        self._thread = thread
+
+    def _atexit(self):
+        self.stop()
+
+    def _atfork_before(self):
+        import time
+        self._lock.acquire()
+        self._atfork_was_stopped = self._stopped
+        thread_was_running = self._thread is not None
+        # Make sure we don't have a running thread
+        while self._thread is not None:
+            thread = self._stop_unlocked()
+            self._lock.release()
+            try:
+                thread.join()
+            finally:
+                self._lock.acquire()
+        if thread_was_running:
+            # HACK: os.fork() queries the system for the number of running threads.
+            # However, Thread.join() only ensures that the Python thread state
+            # was destroyed, while the system thread could still be running.
+            # Give it time to exit.
+            time.sleep(0.001)
+
+    def _atfork_after_in_parent(self):
+        self._stopped = self._atfork_was_stopped
+        self._lock.release()
+
+    def _atfork_after_in_child(self):
+        self._lock = threading.RLock()
+        self._queue = queue.SimpleQueue()
+
+    @property
+    def stopped(self):
+        with self._lock():
+            return self._stopped
+
+    def _stop_unlocked(self):
+        thread = self._thread
+        self._stopped = True
+        if self._thread is not None:
+            self._queue.put(self._sentinel)
+            self._thread = None
+        return thread
+
+    def stop(self):
+        with self._lock:
+            thread = self._stop_unlocked()
+        # To avoid deadlocks between self._lock and the thread-state lock,
+        # call `thread.join` unlocked.
+        if thread is not None:
+            thread.join()
+
+    def reset(self):
+        # For tests
+        self.stop()
+        assert self._queue.qsize() == 0
+        self._stopped = False
+
+    def wait_until_idle(self):
+        with self._queue_drained_cond:
+            self._queue_drained_cond.wait_for(lambda: self._queue.empty())
+
+    def enqueue_task(self, cb):
+        # Can be run concurrently or reentrantly along stop(), or even
+        # from the work loop.
+        assert callable(cb)
+        with self._lock:
+            if not self._stopped and self._thread is not None:
+                # Even if called reentrantly from stop(), the fact that
+                # self._stopped is False ensures that our callback will
+                # be enqueued before the sentinel.
+                self._queue.put(cb)
+                return
+        # The work loop is not running (perhaps we're shutting down?),
+        # execute callback directly without taking the lock.
+        # This might cause issues if this is a reentrant call, but we cannot
+        # do any better.
+        try:
+            cb()
+        finally:
+            cb = None
```

Comment:
Maybe add a reference to `_work_loop()` or copy the comment? I was confused when I saw this.

Hmm, `_work_queue.enqueue_task()` is always called from `the global `enqueue_task()` which keeps a reference to `cb`, so it perhaps does not help.